README.md
![](https://cdn.rawgit.com/hashicorp/terraform-website/master/content/source/assets/images/logo-hashicorp.svg)

# Terraform Provider CIDR

Available in the [Terraform Registry](https://registry.terraform.io/providers/volcano-coffee-company/cidr/latest).

## Requirements

-	[Terraform](https://www.terraform.io/downloads.html) >= 0.13.x
-	[Go](https://golang.org/doc/install) >= 1.15

## Building The Provider

1. Clone the repository
1. Enter the repository directory
1. Build the provider using the Go `install` command: 
```sh
$ go install
```

## Adding Dependencies

This provider uses [Go modules](https://github.com/golang/go/wiki/Modules).
Please see the Go documentation for the most up to date information about using Go modules.

To add a new dependency `github.com/author/dependency` to your Terraform provider:

```
go get github.com/author/dependency
go mod tidy
```

Then commit the changes to `go.mod` and `go.sum`.

## Using the provider

```terraform
data "cidr_network" "example1" {
  prefix = "192.168.2.56/29"
}

data "cidr_network" "example2" {
  ip   = "192.168.2.57"
  mask = "255.255.255.248"
}
```

## Developing the Provider

If you wish to work on the provider, you'll first need [Go](http://www.golang.org) installed on your machine (see [Requirements](#requirements) above).

To compile the provider, run `go install`. This will build the provider and put the provider binary in the `$GOPATH/bin` directory.

To generate or update documentation, run `go generate`.

In order to run the full suite of Acceptance tests, run `make testacc`.

*Note:* Acceptance tests create real resources, and often cost money to run.

```sh
$ make testacc
```


CHANGELOG.md
## 0.1.0 (April 03, 2021)

- Initial release


internal/provider/provider_test.go
package provider

import (
	"testing"

	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
)

// providerFactories are used to instantiate a provider during acceptance testing.
// The factory function will be invoked for every Terraform CLI command executed
// to create a provider server to which the CLI can reattach.
var providerFactories = map[string]func() (*schema.Provider, error){
	"cidr": func() (*schema.Provider, error) {
		return New(), nil
	},
}

func TestProvider(t *testing.T) {
	if err := New().InternalValidate(); err != nil {
		t.Fatalf("err: %s", err)
	}
}

func testAccPreCheck(t *testing.T) {
	// You can add code here to run prior to any test case execution, for example assertions
	// about the appropriate environment variables being set are common to see in a pre-check
	// function.
}


internal/provider/data_source_network_test.go
package provider

import (
	"fmt"
	"regexp"
	"testing"

	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
)

func TestAccDataSourceNetwork_IPv4Mask(t *testing.T) {
	resource.UnitTest(t, resource.TestCase{
		PreCheck:          func() { testAccPreCheck(t) },
		ProviderFactories: providerFactories,
		Steps: []resource.TestStep{
			{
				Config: testAccDataSourceNetworkIPMask("192.168.0.1", "255.255.255.0"),
				Check: resource.ComposeTestCheckFunc(
					resource.TestCheckResourceAttr("data.cidr_network.test", "prefix", "192.168.0.1/24"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "ip", "192.168.0.1"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask", "255.255.255.0"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "network", "192.168.0.0/24"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "first_ip", "192.168.0.1"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "last_ip", "192.168.0.254"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "broadcast", "192.168.0.255"),
				),
			},
		},
	})
}

func TestAccDataSourceNetwork_IPv4Prefix(t *testing.T) {
	resource.UnitTest(t, resource.TestCase{
		PreCheck:          func() { testAccPreCheck(t) },
		ProviderFactories: providerFactories,
		Steps: []resource.TestStep{
			{
				Config: testAccDataSourceNetworkPrefix("192.168.0.1/22"),
				Check: resource.ComposeTestCheckFunc(
					resource.TestCheckResourceAttr("data.cidr_network.test", "prefix", "192.168.0.1/22"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "network", "192.168.0.0/22"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "ip", "192.168.0.1"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask", "255.255.252.0"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask_bits", "22"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "first_ip", "192.168.0.1"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "last_ip", "192.168.3.254"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "broadcast", "192.168.3.255"),
				),
			},
			{
				Config: testAccDataSourceNetworkPrefix("192.168.0.1/24"),
				Check: resource.ComposeTestCheckFunc(
					resource.TestCheckResourceAttr("data.cidr_network.test", "prefix", "192.168.0.1/24"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "network", "192.168.0.0/24"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "ip", "192.168.0.1"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask", "255.255.255.0"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask_bits", "24"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "first_ip", "192.168.0.1"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "last_ip", "192.168.0.254"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "broadcast", "192.168.0.255"),
				),
			},
			{
				Config: testAccDataSourceNetworkPrefix("192.168.0.1/30"),
				Check: resource.ComposeTestCheckFunc(
					resource.TestCheckResourceAttr("data.cidr_network.test", "prefix", "192.168.0.1/30"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "network", "192.168.0.0/30"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "ip", "192.168.0.1"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask", "255.255.255.252"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask_bits", "30"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "first_ip", "192.168.0.1"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "last_ip", "192.168.0.2"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "broadcast", "192.168.0.3"),
				),
			},
			{
				Config: testAccDataSourceNetworkPrefix("192.168.0.1/31"),
				Check: resource.ComposeTestCheckFunc(
					resource.TestCheckResourceAttr("data.cidr_network.test", "prefix", "192.168.0.1/31"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "network", "192.168.0.0/31"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "ip", "192.168.0.1"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask", "255.255.255.254"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask_bits", "31"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "first_ip", "192.168.0.0"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "last_ip", "192.168.0.1"),
				),
			},
			{
				Config: testAccDataSourceNetworkPrefix("192.168.0.1/32"),
				Check: resource.ComposeTestCheckFunc(
					resource.TestCheckResourceAttr("data.cidr_network.test", "prefix", "192.168.0.1/32"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "network", "192.168.0.1/32"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "ip", "192.168.0.1"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask", "255.255.255.255"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask_bits", "32"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "first_ip", "192.168.0.1"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "last_ip", "192.168.0.1"),
				),
			},
		},
	})
}

func TestAccDataSourceNetwork_IPv6Mask(t *testing.T) {
	resource.UnitTest(t, resource.TestCase{
		PreCheck:          func() { testAccPreCheck(t) },
		ProviderFactories: providerFactories,
		Steps: []resource.TestStep{
			{
				Config: testAccDataSourceNetworkIPMask("2a03:2880:20:4f06::", "ffff:ffff:ffff:ffff::"),
				Check: resource.ComposeTestCheckFunc(
					resource.TestCheckResourceAttr("data.cidr_network.test", "prefix", "2a03:2880:20:4f06::/64"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "ip", "2a03:2880:20:4f06::"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask", "ffff:ffff:ffff:ffff::"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask_bits", "64"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "network", "2a03:2880:20:4f06::/64"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "first_ip", "2a03:2880:20:4f06::"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "last_ip", "2a03:2880:20:4f06:ffff:ffff:ffff:ffff"),
				),
			},
		},
	})
}

func TestAccDataSourceNetwork_IPv6Prefix(t *testing.T) {
	resource.UnitTest(t, resource.TestCase{
		PreCheck:          func() { testAccPreCheck(t) },
		ProviderFactories: providerFactories,
		Steps: []resource.TestStep{
			{
				Config: testAccDataSourceNetworkPrefix("2a03:2880:20:4f06::/64"),
				Check: resource.ComposeTestCheckFunc(
					resource.TestCheckResourceAttr("data.cidr_network.test", "prefix", "2a03:2880:20:4f06::/64"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "ip", "2a03:2880:20:4f06::"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask", "ffff:ffff:ffff:ffff::"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask_bits", "64"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "network", "2a03:2880:20:4f06::/64"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "first_ip", "2a03:2880:20:4f06::"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "last_ip", "2a03:2880:20:4f06:ffff:ffff:ffff:ffff"),
				),
			},
			{
				Config: testAccDataSourceNetworkPrefix("1::/64"),
				Check: resource.ComposeTestCheckFunc(
					resource.TestCheckResourceAttr("data.cidr_network.test", "prefix", "1::/64"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "ip", "1::"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask", "ffff:ffff:ffff:ffff::"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask_bits", "64"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "network", "1::/64"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "first_ip", "1::"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "last_ip", "1::ffff:ffff:ffff:ffff"),
				),
			},
			{
				Config: testAccDataSourceNetworkPrefix("::ffff:192.168.0.0/112"),
				Check: resource.ComposeTestCheckFunc(
					resource.TestCheckResourceAttr("data.cidr_network.test", "prefix", "192.168.0.0/112"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "ip", "192.168.0.0"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask", "ffff:ffff:ffff:ffff:ffff:ffff:ffff:0"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask_bits", "112"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "network", "192.168.0.0/16"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "first_ip", "192.168.0.0"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "last_ip", "192.168.255.255"),
				),
			},
			{
				Config: testAccDataSourceNetworkPrefix("fe80::/48"),
				Check: resource.ComposeTestCheckFunc(
					resource.TestCheckResourceAttr("data.cidr_network.test", "prefix", "fe80::/48"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "ip", "fe80::"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask", "ffff:ffff:ffff::"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask_bits", "48"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "network", "fe80::/48"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "first_ip", "fe80::"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "last_ip", "fe80::ffff:ffff:ffff:ffff:ffff"),
				),
			},
			{
				Config: testAccDataSourceNetworkPrefix("fe80::3:0:0:0/81"),
				Check: resource.ComposeTestCheckFunc(
					resource.TestCheckResourceAttr("data.cidr_network.test", "prefix", "fe80::3:0:0:0/81"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "ip", "fe80::3:0:0:0"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask", "ffff:ffff:ffff:ffff:ffff:8000::"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "mask_bits", "81"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "network", "fe80::3:0:0:0/81"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "first_ip", "fe80::3:0:0:0"),
					resource.TestCheckResourceAttr("data.cidr_network.test", "last_ip", "fe80::3:7fff:ffff:ffff"),
				),
			},
		},
	})
}

func TestAccDataSourceNetwork_MissingArgs(t *testing.T) {
	resource.UnitTest(t, resource.TestCase{
		PreCheck:          func() { testAccPreCheck(t) },
		ProviderFactories: providerFactories,
		Steps: []resource.TestStep{
			{
				Config:      testAccDataSourceNetworkMissingArgs,
				ExpectError: regexp.MustCompile("one of `ip,prefix` must be specified"),
			},
		},
	})
}

func TestAccDataSourceNetwork_MissingIP(t *testing.T) {
	resource.UnitTest(t, resource.TestCase{
		PreCheck:          func() { testAccPreCheck(t) },
		ProviderFactories: providerFactories,
		Steps: []resource.TestStep{
			{
				Config:      testAccDataSourceNetworkMissingIP,
				ExpectError: regexp.MustCompile("one of `ip,prefix` must be specified"),
			},
		},
	})
}

func TestAccDataSourceNetwork_MissingMask(t *testing.T) {
	resource.UnitTest(t, resource.TestCase{
		PreCheck:          func() { testAccPreCheck(t) },
		ProviderFactories: providerFactories,
		Steps: []resource.TestStep{
			{
				Config:      testAccDataSourceNetworkMissingMask,
				ExpectError: regexp.MustCompile("all of `ip,mask` must be specified"),
			},
		},
	})
}

func testAccDataSourceNetworkPrefix(prefix string) string {
	return fmt.Sprintf(`
data "cidr_network" "test" {
  prefix = "%s"
}
`, prefix)
}

func testAccDataSourceNetworkIPMask(ip, mask string) string {
	return fmt.Sprintf(`
data "cidr_network" "test" {
  ip   = "%s"
  mask = "%s"
}
`, ip, mask)
}

const testAccDataSourceNetworkMissingArgs = `
data "cidr_network" "test" {
}
`

const testAccDataSourceNetworkMissingIP = `
data "cidr_network" "test" {
  mask = "255.255.255.0"
}
`

const testAccDataSourceNetworkMissingMask = `
data "cidr_network" "test" {
  ip = "192.168.2.56"
}
`


main.go
package main

import (
	"context"
	"flag"
	"log"

	"github.com/hashicorp/terraform-plugin-sdk/v2/plugin"
	"github.com/volcano-coffee-company/terraform-provider-cidr/internal/provider"
)

// Run "go generate" to format example terraform files and generate the docs for the registry/website

// If you do not have terraform installed, you can remove the formatting command, but its suggested to
// ensure the documentation is formatted properly.
//go:generate terraform fmt -recursive ./examples/

// Run the docs generation tool, check its repository for more information on how it works and how docs
// can be customized.
//go:generate go run github.com/hashicorp/terraform-plugin-docs/cmd/tfplugindocs

func main() {
	var debugMode bool

	flag.BoolVar(&debugMode, "debug", false, "set to true to run the provider with support for debuggers like delve")
	flag.Parse()

	opts := &plugin.ServeOpts{ProviderFunc: provider.New}

	if debugMode {
		err := plugin.Debug(context.Background(), "registry.terraform.io/volcano-coffee-company/cidr", opts)
		if err != nil {
			log.Fatal(err.Error())
		}
		return
	}

	plugin.Serve(opts)
}


internal/provider/data_source_network.go
package provider

import (
	"context"
	"fmt"
	"net"

	"github.com/apparentlymart/go-cidr/cidr"
	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
)

func dataSourceNetwork() *schema.Resource {
	return &schema.Resource{
		Description: "Use this data source to manipulate network addresses, such as IPv4 and IPv6 addresses, subnets, masks and prefixes.",

		ReadContext: dataSourceNetworkRead,

		Schema: map[string]*schema.Schema{
			"prefix": {
				Description: "The IPv4 or IPv6 address and network mask in CIDR notation. " +
					"The address may denote a specific address (using a host identifier, such as `10.0.0.1/8`), " +
					"or the beginning address of an entire network (using a host identifier of 0, such as `10.0.0.0/8`).",
				Type:             schema.TypeString,
				Optional:         true,
				Computed:         true,
				ExactlyOneOf:     []string{"prefix", "ip"},
				ValidateDiagFunc: validation.ToDiagFunc(validation.IsCIDR),
			},
			"ip": {
				Description:      "The IPv4 or IPv6 address (such as `10.0.0.1` or `2001:db8::68`).",
				Type:             schema.TypeString,
				Optional:         true,
				Computed:         true,
				RequiredWith:     []string{"ip", "mask"},
				ValidateDiagFunc: validation.ToDiagFunc(validation.IsIPAddress),
			},
			"mask": {
				Description:      "The IPv4 or IPv6 network mask (such as `255.0.0.0` or `ffff:ffff:ffff::`).",
				Type:             schema.TypeString,
				Optional:         true,
				Computed:         true,
				ValidateDiagFunc: validation.ToDiagFunc(validation.StringIsNotEmpty),
			},
			"mask_bits": {
				Description: "The number of bits in the IPv4 or IPv6 network mask.",
				Type:        schema.TypeInt,
				Computed:    true,
			},
			"network": {
				Description: "The IPv4 or IPv6 network in CIDR notation (such as `10.0.0.0/8` or `2001:db8::/48`).",
				Type:        schema.TypeString,
				Computed:    true,
			},
			"first_ip": {
				Description: "The first IPv4 or IPv6 address in the CIDR range (such as `10.0.0.1` or `2001:db8::`).",
				Type:        schema.TypeString,
				Computed:    true,
			},
			"last_ip": {
				Description: "The last IPv4 or IPv6 address in the CIDR range (such as `10.255.255.254` or `2001:db8:0:ffff:ffff:ffff:ffff:ffff`).",
				Type:        schema.TypeString,
				Computed:    true,
			},
			"broadcast": {
				Description: "The IPv4 broadcast address (such as `10.255.255.255`).",
				Type:        schema.TypeString,
				Computed:    true,
			},
		},
	}
}

func dataSourceNetworkRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
	var err error
	var broadcast net.IP
	var ip net.IP
	var mask net.IPMask
	var network *net.IPNet

	if v, ok := d.GetOk("prefix"); ok {
		ip, network, err = net.ParseCIDR(v.(string))
		if err != nil {
			return diag.FromErr(err)
		}
	}

	if v, ok := d.GetOk("ip"); ok {
		ip = net.ParseIP(v.(string))
	}

	if v, ok := d.GetOk("mask"); ok {
		m := net.ParseIP(v.(string))
		switch {
		case len(m.To4()) == net.IPv4len:
			mask = net.IPv4Mask(m[12], m[13], m[14], m[15])
		case len(m) == net.IPv6len:
			mask = net.IPMask(m)
		}

		network = &net.IPNet{
			IP:   ip.Mask(mask),
			Mask: mask,
		}
	}

	firstIP, lastIP := cidr.AddressRange(network)

	ones, _ := network.Mask.Size()
	if network.IP.To4() != nil {
		if ones < 31 {
			firstIP = cidr.Inc(firstIP)
			lastIP = cidr.Dec(lastIP)
		}

		broadcast = make(net.IP, len(network.IP.To4()))
		for i := 0; i < len(network.IP.To4()); i++ {
			broadcast[i] = network.IP[i] | ^network.Mask[i]
		}
	}

	d.Set("prefix", fmt.Sprintf("%s/%d", ip.String(), ones))
	d.Set("ip", ip.String())
	d.Set("mask", net.IP(network.Mask).String())
	d.Set("mask_bits", ones)
	d.Set("network", network.String())
	d.Set("first_ip", firstIP.String())
	d.Set("last_ip", lastIP.String())
	d.Set("broadcast", broadcast.String())

	d.SetId(network.String())

	return nil
}
